### ISSUE TITLE
[css-shapes][css3-backgrounds] No explicit <position> serialization steps in background spec

### ISSUE THREAD

**Issue by Manishearth** [w3c/csswg-drafts#368]
https://drafts.csswg.org/css-shapes/#basic-shape-serialization

https://drafts.csswg.org/css-backgrounds-3/#position

The basic-shape spec mentions:

> The <position> values (including the defaults) in ellipse() and circle() serialize to their 2- and 4-value forms only, preferring the 2-value form when it can be expressed without calc(), preferring left and top origins, and preferring 0% over a zero length.

The backgrounds spec does not have serialization steps mentioned at all. Can we assume the same serialization steps?

(This doesn't match existing behavior -- browsers serialize `center center` as `center center` instead of `50% 50%` for background position)

**Comment by upsuper** [comment]
I don't think we want to change that behavior. It seems to me functions usually have special serialization rules, but not general property values.

**Comment by Manishearth** [comment]
Does this mean we can serialize any way we want provided that it parses to the same thing?

**Comment by birtles** [comment]
I don't think so. For one thing it makes writing web-platform-tests hard (not to mention making life hard for scripts that want to parse the result of `getComputedStyle`). We recently specced &lt;timing-function&gt; serialization (and fixed it in Gecko so far, Blink later) so we can write interop tests. We should probably spec `background-position` / `<position>` too.

**Comment by upsuper** [comment]
It doesn't seem to me serialization of general specified value is defined anywhere... If we are trying to define it, it should match what engines are currently doing.

My guess is that engines general just put the parsed tokens in the grammar order, so `top center` => `center top`.

**Comment by upsuper** [comment]
This should probably be defined in css-values spec I suppose.

**Comment by birtles** [comment]
There's a fair bit of serialization defined in CSSOM, e.g. [serializing a CSS value](https://drafts.csswg.org/cssom/#serializing-css-values). But I don't think it covers `<position>`.

**Comment by cdoublev** [comment]
I was looking for issues related to the serialization of `<position>` (ie. the "generic" `<position>`) and `<bg-position>` because I do not understand the rationale for the following (quoted from [CSS Shapes, 3.3 Serialization of Basic Shapes](https://drafts.csswg.org/css-shapes-1/#basic-shape-serialization)), which is matching the current browser outputs and [WPT tests](https://github.com/web-platform-tests/wpt/blob/3c77bf496e9c506b6ce32cf5866fa3a97f95f881/css/css-shapes/basic-shape-circle-ellipse-serialization.html#L27):

> Omitting components means that some default values do not show up in the serialization. But since <position> always uses the 2- or 4-value form, a default <position> is not omitted.
>
> `circle(closest-side at center)` serializes as `circle(at 50% 50%)`

I guess that *<position> always uses the 2- or 4-value form* comes from an old version of CSS Shape, but why this rule is also applied to `<bg-position>`? And why would it be required? Furthermore, the following (quoted from [CSSOM, 6.7.2 Serializing CSS Values](https://drafts.csswg.org/cssom/#serializing-css-values)) is a normative text against it:

> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.

`circle()`, `circle(at center)`, `circle(at 50%)`, ... are all valid and *means the same value*, and my guess would have been to serialize them as `circle()`. I am not claiming that they should be serialized like this, but I just woud like to understand the current browser outputs.

**Comment by astearns** [comment]
@cdoublev (and others) take a look at the discussion in https://github.com/w3c/csswg-drafts/issues/2274 for some of the things that led us to decide to only serialize to 2- and 4-value forms. The text in css-shapes was also supposed to also go in css-backgrounds-3 (I remember having plans to remove as much as I could from css-shapes and just refer to bg-position serialization when I could)

**Comment by cdoublev** [comment]
Thanks, I would have found this issue if I had searched on this topic without the `css-shapes` label.

If I understood correctly :
- `<bg-position>` should be serialized as `<position>`
- to sum up the two following points: `<position>` should be serialized with 2 or 4 values
- `<position>` should be serialized with all components of the syntax matched by the input value, no matter if they were omitted or equal or equivalent to the default value (eg. `50%`, `center`, `center 50%`, `50% center`)¹, because it would complicate round-tripping (from/to parsed/serialized values)², and because parsing `transform-origin` would become ambiguous
- `<position>` specified with 3 values ​​must be serialized by adding a missing keyword (`top` or` left`), or by adding a missing percentage and eventually replacing the keyword (`bottom` by ` top `,` right` by `left`)

The [current spec for `transform-origin`](https://drafts.csswg.org/css-transforms/#propdef-transform-origin) does not use `<position>`, but I'm assuming that it may be updated to specify that `transform-origin` should be serialized as `<position>`.

¹ I should note that it makes sens to me also because it would be weird to get `circle(center 50px)` not serialized to `circle(50px)` when the target box is 100px tall, and because an exception would have to exist for keeping at least 1 value when `<position>` is used alone to define the property.
² I understand this point even though I think it might not be not so obvious to a CSS developer.

**Comment by cdoublev** [comment]
And I guess the reasoning for `<position>` does not apply to `<border-radius>` in `inset(<length-percentage>{1,4} [round <'border-radius'>]?)`.

| `clip-path`                | Chrome 96             | Firefox 94   |
| -------------------------- | --------------------- | ------------ | 
| `inset(1px)`               | `inset(1px)`          | `inset(1px)` |
| `inset(1px round 0)`       | `inset(1px)`          | `inset(1px)` |
| `inset(1px round calc(0))` | `inset(1px)`          | `inset(1px)` |
| `inset(1px round 0%)`      | `inset(1px round 0%)` | `inset(1px)` |
| `inset(1px round 0% / 0%)` | `inset(1px round 0%)` | `inset(1px)` |

| `border-radius` | Chrome 96  | Firefox 94 |
| --------------- | ---------- | ---------- | 
| `0`             | `0px`      | `0px`      |
| `0%`            | `0%`       | `0%`       |

**Comment by fantasai** [comment]
This is now edited into css-values-4, see https://github.com/w3c/csswg-drafts/commit/0742ad6e372782b46cca9708fa293e2b5abab219

**Comment by cdoublev** [comment]
The issue seems also about missing serialization steps for `background-position` in CSS Backgrounds, noting that CSS Backgrounds 3 uses `<bg-position>`, and CSS Backgrounds 4 uses `<position>` ([related](https://github.com/w3c/csswg-drafts/issues/7376)).

**Comment by SebastianZ** [comment]
For reference, https://github.com/w3c/csswg-drafts/commit/ed78c136918500a6fc0995869363b0feea11fdae defines serialization steps for CSS Backgrounds 3 (and [got already incorported into level 4](https://github.com/w3c/csswg-drafts/commit/d45bed77b7a055a54c0b204db2c0e4cdc7628b06)).

Sebastian
